### PR TITLE
Protein-Protein XLMS Part I: writing and reading XLMS data in MzIdentML format

### PR DESCRIPTION
**Pull request by enetz**
The XLMOD.obo was added and is read into the ModificationsDB.
Fragment annotations were added to PeptideHit.
Test and small test files for XLMS data were added.

The search engine using this will be added with another pull request later.

### CODE REVIEW

**Comment by enetz** [comment]
Most of the failed tests are because of the changes to ModificationsDB.
I tried to skip reading in the XLMOD.obo into the ModsDB, but that only solves 3 of 9 errors for the PepXMLFile_test. The changes to ModsDB in this PR also solve a few errors in reading in the PSI-MOD.obo, so the number and order of modifications in the DB is changed.
The failed tests probably happen because ambiguities are solved by taking the first mod in a vector of found mods, but additional modifications read in from PSI-MOD mess with their order.
Are the choices in the tests based on a known ground truth or is simply the first option chosen each time?
Would updating the "correct" choice make sense?

**Comment by hendrikweisser** [comment]
> The failed tests probably happen because ambiguities are solved by taking the first mod in a vector of found mods, but additional modifications read in from PSI-MOD mess with their order.

In case of ambiguities (e.g. when matching by mass) taking the first mod from Unimod is usually a good heuristic, because the more "common" mods tend to appear early. If you're adding new modifications, can't you make sure they're added a the end of the relevant list, and shouldn't that solve the problem?

**Comment by enetz** [comment]
Well I am doing that already, since the function readFromOBOFile() uses "mods_.push_back(new_mod)", so everything from unimod comes before PSI-MOD and PSI-MOD comes before XLMOD in the mods_ vector. The problem seems to be the "search..." and "get..." functions.
Some of them use std::set, which orders stuff differently.
In that case it does not matter in which order they are added.

**Comment by enetz** [comment]
I am currently working on a workaround specifically for cross-links:
Since I do not expect any situation in which someone would search for normal modifications and cross-links at the same time, I am writing a CrossLinksDB class which inherits from ModificationsDB but reads in only the XLMOD.obo. 
That way we would avoid finding cross-links instead of normal mods in ModsDB.
I can also revert the "readFromOBOFile()" function to its former state and use the new version in the CrossLinksDB, meaning some mods will be missing but ModsDB will function as it was. Nobody missed these mods until now (specifically all PSI-MOD mods with an Origin string longer than 1, like "D, E").

**Comment by timosachsenberg** [comment]
@enetz would it be possible to add a search for both cross-links and normal modifications? I could imagine that there might be some applications where this would make sense.

**Comment by enetz** [comment]
It would make sense for mono-links, which can be treated as normal modifications. They are read in with the cross-links right now, but would actually belong in ModificationsDB. I am unsure about how to treat this, since even a few additional mods from PSI-MOD already broke a few tests. Maybe we can prioritize UniMod mods somehow. We could keep two vectors of mods in ModificationsDB and only search in the second one, if nothing was found in the first containing UniMod. I think this would be straightforward to implement.
Then we could read in the mods from PSI-MOD and mono-links from XLMOD into the second vector.

**Comment by enetz** [comment]
@mwalzer I am running into issues because of old tests. 
A few are due to general fixes and updates and I am updating the tested files,
but I want to make sure I don't introduce errors into the tests.

For example you updated the fragmentation table in the ppxlink branch (quite sure this is a feature) from:
> 	<Measure id="Measure_SIL_4835329514588776807">
> 		<cvParam accession="MS:1001225" ...  name="product ion m/z"/>
> 	</Measure>
to:
> 	<Measure id="Measure_mz">
> 		<cvParam accession="MS:1001225" ... name="product ion m/z"/>
> 	</Measure>
> 	<Measure id="Measure_int">
> 		<cvParam accession="MS:1001226" ... unitName="number of detector counts"/>
> 	</Measure>
> 	<Measure id="Measure_error">
> 		<cvParam accession="MS:1001227" ... unitName="m/z"/>
> 	</Measure>

Is this next one also a feature or a bug?
Scores in SpectrumIdentificationItem changed from (cvParam with value):
> <cvParam accession="MS:1001143" cvRef="PSI-MS" name="search engine specific score for PSMs" value="1.4"/>
to (cvParam without value + userParam with value):
> <cvParam accession="MS:1001143" cvRef="PSI-MS" name="search engine specific score for PSMs"/>
> <userParam name="MOWSE" unitName="xsd:double" value="1.4"/>

Another one are multiple PeptideEvidence lines in SII:
There is one peptide contained in two different proteins, so it has two PeptideEvidences in the sequence collection.
In the test file the SII with this Peptide only has one PeptideEvidenceRef, but now two are written.
I think there should be two, since from the SII you can not know to which of the two Proteins this peptide belongs, or more precisely it belongs to both.

**Comment by mwalzer** [comment]
@enetz yes, these got introduced with better cv handling in mzid for PeptideIdentification

File: src/openms/source/CHEMISTRY/Residue.cpp [hunk 1]
```diff
@@ -508,6 +508,12 @@ namespace OpenMS
     {
       mono_weight_ = mod.getMonoMass();
     }
+    // update mono_weight_ by DiffMonoMass, if MonoMass is not known, but DIffMonoMass is
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Residue.cpp]: DiffMonoMass

**Comment by timosachsenberg** [comment]
hi @enetz could you add a small summary what changed in the core data structures like PeptideHit so people get quick overview.

**Comment by enetz** [comment]
ModificationsDB:
function "readFromOBOFile" was expanded, it reads in more mods from PSI-MOD.obo and the mono-links from XLMOD.obo. Unfortunately these additional mods change the order of mods, when searching by mass, so in some cases a different mod might be chosen (or just a different name for the same molecular structure). This happened in the tests with reading in pepXML files without mod names. For now I solved this by commenting out the conflicting mods in PSI-MOD.obo.
Also an additional class CrossLinksDB was added. It is a friend class of ModificationsDB and reads in only the cross-links from XLMOD.obo. Otherwise it inherits the same functionality as ModsDB.

**Comment by mwalzer** [comment]
Thanks, could you also include the core data structures changes please?

**Comment by enetz** [comment]
PeptideHit:
the new struct FragmentAnnotation was added. It stores the intensity, charge and position of one MS peak, as well as an annotation string.
A new member fragment_annotations_, a vector of these structs, was also added to the class, including a getter and setter function.

**Comment by enetz** [comment]
MzIdentMLHandler (writing mzid) and MzIdentMLDOMHandler (reading mzid) were expanded to write and read XL-MS data according to MzIdentML 1.2 specifications with additional fragment annotation data.
Some fixes and improved CV param handling are also included, e.g. correct reading and writing of peptide start and end positions, a correction in parsing CVTerm::Unit objects, writing an improved fragmentation table and a few other improvements.

**Comment by timosachsenberg** [comment]
looks good but tests are required for the core classes and ideally also python bindings

**Comment by hroest** [comment]
seems like a lot of the tests are failing and I agree with @timosachsenberg that we should include tests and Python bindings

**Comment by timosachsenberg** [comment]
looks good to me. Thanks for going through the pain of working on core data structures and classes ;) If problems show up with this PR you can fix them in the other PRs.

File: src/openms/include/OpenMS/CHEMISTRY/CrossLinksDB.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Timo Sachsenberg $
+// $Authors: Andreas Bertsch $
+// --------------------------------------------------------------------------
+
+#ifndef CROSSLINKSDB_H
+#define CROSSLINKSDB_H
```
Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/CrossLinksDB.h]: I think this needs to be changed - that is not according to OpenMS coding style

File: src/openms/include/OpenMS/METADATA/PeptideHit.h [hunk 3]
```diff
@@ -57,6 +57,64 @@ namespace OpenMS
   {
 public:
 
+  struct FragmentAnnotation
+  {
+    String annotation;  // e.g. [alpha|ci$y3-H2O-NH3]5+
```
Comment (hroest) [inline src/openms/include/OpenMS/METADATA/PeptideHit.h]: wow, what magic syntax is this? could this be documented somewhere what these individual symbols mean?

**Pull request merged**